### ISSUE TITLE
%Matlab Magic With Native Matlab Engine?

### ISSUE THREAD

**Issue by Foucl** [Calysto/matlab_kernel#60]
As far as I know, you can only create matlab cells in other kernels through pymatbridge's  `(%)%matlab` magic. However, I can only reliablly get the python-matlab connection to work through your kernel and the native matlab python engine. It would be so great if I were able to define matlab cells in my python projects to connect to Matlab's python engine. Can this be done using your code here and/or meta_kernel?

**Comment by dsblank** [comment]
If I understand you correctly, you want to expose the matlab_kernel to IPython kernels? Yes, that should be possible and easy. For example, see how the %%scheme magic works:

https://github.com/Calysto/metakernel/blob/master/metakernel/magics/scheme_magic.py

especially the register_ipython_magics at bottom. This is used in ipython by installing calysto_scheme and metakernel, and then in an ipython notebook:

```python
import metakernel; metakernel.register_ipython_magics()
```
Then %%scheme will work in ipython.

I don't have matlab, so I can't really test this, but should be straightforward to write such a magic.

**Comment by blink1073** [comment]
What Doug said :wink:  I also do not have a working Matlab, but if you submit a PR that is working for you @Foucl , I will accept it.

**Comment by dsblank** [comment]
Oh, actually I see we have already done something that will work. This is the generic kernel magic exposed in IPython:

First register the kernel and kx magics:

```python
import metakernel; metakernel.register_ipython_magics()
```

Next, start up a MatlabKernel

```python
%kernel matlab_kernel.kernel MatlabKernel
```

And now send it commands:

```python
%%kx 
... matlab commands
```

**Comment by Foucl** [comment]
That is exactly what I am looking for! However, trying to do this in ipython i can create the kernel (with output `No handlers could be found for logger ".metakernel"`), but executing commands does not work:
```
%%kx
a = 2 
```
gives me
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-865d3c18e798> in <module>()
----> 1 get_ipython().run_cell_magic(u'kx', u'', u'a = 2')

c:\users\ddmitarbeiter\anaconda2\lib\site-packages\IPython\core\interactiveshell.pyc in run_cell_magic(self, magic_name, line, cell)
   2113             magic_arg_s = self.var_expand(line, stack_depth)
   2114             with self.builtin_trap:
-> 2115                 result = fn(magic_arg_s, cell)
   2116             return result
   2117 

c:\users\ddmitarbeiter\anaconda2\lib\site-packages\metakernel\magics\kernel_magic.pyc in kx(line, cell)
    123             module_name = "default"
    124         magic.code = cell
--> 125         magic.cell_kx(module_name)

c:\users\ddmitarbeiter\anaconda2\lib\site-packages\metakernel\magics\kernel_magic.pyc in cell_kx(self, kernel_name)
     61         if kernel_name is None:
     62             kernel_name = self.kernel_name
---> 63         self.retval = self.kernels[kernel_name].do_execute_direct(self.code)
     64         self.evaluate = False
     65 

c:\users\ddmitarbeiter\anaconda2\lib\site-packages\matlab_kernel\kernel.pyc in do_execute_direct(self, code)
    136             stdout = resp['content']['stdout'].strip()
    137             if stdout:
--> 138                 self.Print(stdout)
    139 
    140     def get_kernel_help_on(self, info, level=0, none_on_fail=False):

c:\users\ddmitarbeiter\anaconda2\lib\site-packages\metakernel\_metakernel.pyc in Print(self, *args, **kwargs)
    623             'name': 'stdout', 'text': message}
    624         self.log.debug('Print: %s' % message)
--> 625         self.send_response(self.iopub_socket, 'stream', stream_content)
    626 
    627     def Write(self, message):

c:\users\ddmitarbeiter\anaconda2\lib\site-packages\ipykernel\kernelbase.pyc in send_response(self, stream, msg_or_type, content, ident, buffers, track, header, metadata)
    344         message.
    345         """
--> 346         return self.session.send(stream, msg_or_type, content, self._parent_header,
    347                                  ident, buffers, track, header, metadata)
    348 

AttributeError: 'NoneType' object has no attribute 'send'
```

**Comment by dsblank** [comment]
You have updated ipython, jupyter, and metakernel?

This works for me:

```
%kernel calysto_scheme CalystoScheme

%%kx 
(print (+ 1 2))
```

**Comment by Foucl** [comment]
**tl;dr**: calysto_scheme works as described by @dsblank. %kernel matlab_kernel still fails with all 3 engines (native, pymatbridge, octave).

Thank you so much for helping me get this to work. ipython and jupyter are the latest version from pip (`4.1.1` and `5.1.0`, respectively). I installed **calysto_scheme** and can get your example to work without issues. By now, I checked all combinations of the conda environment from which I started ipython (python2.7 vs 3.4) and the kernel from which I tried to make the connection to Matlab.

To further narrow down the problem, I installed **Octave** and set the `OCTAVE_EXECUTABLE` to force matlab_kernel to use octave. Opening Octave (both gui and cli) works fine. After manually doing `from pymatbridge import Octave` this gives me the exact same error message (complaining about missing .metakernel logger and then the error quoted in my last message). However, pymatbridge does not seem to support octave on windows (`octex = os.environ.get('OCTAVE_EXECUTABLE'); Octave(octex).start()` causes a windows-popup complaining about missing liboctinerp-3.dll to open).

Finally, I checked with **pymatbridge** (by commenting out [these lines](https://github.com/Calysto/matlab_kernel/blob/master/matlab_kernel/kernel.py#L29-L31) in matlab_kernel/kernel.py. pymatbridge itself (as well as its matlab-magic) works ok. However, as a backend for matlab_kernel, I can no longer get it working at all. on `jupyter console --debug --kernel matlab` I see the pymatbridge's Window opening up in the background as usual. However, the console crashes with some `TypeError: <StringIO.StringIO instance at 0x0000000007A52688> is not JSON serializable` (full traceback in a [gist](https://gist.github.com/Foucl/e948d900e26541459b392d040fcc0ed7)).

**Comment by dsblank** [comment]
This looks like it might be a process_metakernel vs. metakernel issue. Perhaps @blink1073 has some insight on this...

**Comment by dsblank** [comment]
@blink1073 I'm thinking that process_metakernels need to have kernel.makeSubkernelTo(process_kernel, ...)  called to get .session set. Or override kernel.send_response...

**Comment by dsblank** [comment]
I think I am seeing that the problem is in metakernel.subkernel()... I think that this actually needs to copy the fields when the subkernel is a process_kernel.

**Comment by blink1073** [comment]
I can look at this this weekend, rather swamped at the moment.

**Comment by blink1073** [comment]
I will try and get this working from the `Octave` kernel.
